tests/unit/clustering/kmeans_unittest.cc

```diff
@@ -162,3 +165,105 @@ TEST(KMeans, minibatch_training_test)
 	SG_UNREF(features);
 }
 
+TEST(KMeans, random_init_fail)
+{
+	/* Create a gaussian with many points. Move very small number of points far away.
+	 * Random initialization should not (very small probability) assign separate
+	 * cluster to the far away points.
+	 */
+	SGVector<float64_t> mean(2);
+	SGMatrix<float64_t> cov(2,2);
+	cov(0,0)=10.0;
+	cov(0,1)=5.0;
+	cov(1,0)=5.0;
+	cov(1,1)=10.0;
+	mean[0]=1.0;
+	mean[1]=1.0;
+	CGaussianDistribution* gauss=new CGaussianDistribution(mean,cov);
+	
+	SGMatrix<float64_t> samples(2, 10000);
+	samples=gauss->sample(10000);
+	samples(0,9998)+=1000.0;
+	samples(1,9998)+=1000.0;
+	samples(0,9999)+=1000.0;
+	samples(1,9999)+=1000.0;
+
+	CDenseFeatures<float64_t>* features=new CDenseFeatures<float64_t>(samples);
+	SG_REF(features);
+	CEuclideanDistance* distance=new CEuclideanDistance(features, features);
+	CKMeans* clustering=new CKMeans(2, distance);
+
+
+	int32_t pass=0;
+
+	for (int32_t loop=0; loop<10; loop++)
+	{
+		clustering->train(features);
+		CDenseFeatures<float64_t>* learnt_centers=CDenseFeatures<float64_t>::obtain_from_generic(distance->get_lhs());
+		SGMatrix<float64_t> c=learnt_centers->get_feature_matrix();
+		
+		if((c(0,0)>=300.0 && c(1,0)>=300.0) || (c(0,1)>=300.0 && c(1,1)>=300.0)  )
+			pass++;
+
+		SG_UNREF(learnt_centers);
+	}
+	
+	EXPECT_LT(pass, 2);
+	SG_UNREF(clustering);
+	SG_UNREF(features);
+	SG_UNREF(gauss)	
+}
+
+TEST(KMeans, manual_init_pass)
+{
+	/* Create a gaussian with many points. Move very small number of points far away.
+	 * Manual initialization should assign separate cluster to the far away points.
+	 */
+	SGVector<float64_t> mean(2);
+	SGMatrix<float64_t> cov(2,2);
+	cov(0,0)=10.0;
+	cov(0,1)=5.0;
+	cov(1,0)=5.0;
+	cov(1,1)=10.0;
+	mean[0]=1.0;
+	mean[1]=1.0;
+	CGaussianDistribution* gauss=new CGaussianDistribution(mean,cov);
+	
+	SGMatrix<float64_t> initial_centers(2,2);
+	initial_centers(0,0)=1.0;
+	initial_centers(1,0)=1.0;
+	initial_centers(0,1)=1000.0;
+	initial_centers(1,1)=1000.0;
+
+	SGMatrix<float64_t> samples(2, 10000);
+	samples=gauss->sample(10000);
+	samples(0,9998)+=1000.0;
+	samples(1,9998)+=1000.0;
+	samples(0,9999)+=1000.0;
+	samples(1,9999)+=1000.0;
+
+	CDenseFeatures<float64_t>* features=new CDenseFeatures<float64_t>(samples);
+	SG_REF(features);
+	CEuclideanDistance* distance=new CEuclideanDistance(features, features);
+	CKMeans* clustering=new CKMeans(2, distance, initial_centers);
+
+
+	int32_t pass=0;
+
+	for (int32_t loop=0; loop<10; loop++)
+	{
+		clustering->train(features);
+		CDenseFeatures<float64_t>* learnt_centers=CDenseFeatures<float64_t>::obtain_from_generic(distance->get_lhs());
+		SGMatrix<float64_t> c=learnt_centers->get_feature_matrix();
+		
+		if((c(0,0)>=300.0 && c(1,0)>=300.0) || (c(0,1)>=300.0 && c(1,1)>=300.0)  )
+			pass++;
+
+		SG_UNREF(learnt_centers);
+	}
+	
+	EXPECT_EQ(pass, 10);
+	SG_UNREF(clustering);
+	SG_UNREF(features);
+	SG_UNREF(gauss)	
+}
```

Comment:
Unit tests are a quite different idea. These here are too big and complicated, if they fail, they dont tell more of what has failed than an integration test that fails due to a changed result

You should test:
 * corner cases
 * on very small datasets
 * Random initialisation: just test that every center is *exactly* on one data point, and no two centers are the same. How many points and k do you need to test that? Try two cases: k=2, n=3   and k=3, n=3
 * test modules or the code, not chains of multiple modules
 * Dont use random data, unless it leads to *guaranteed* pass